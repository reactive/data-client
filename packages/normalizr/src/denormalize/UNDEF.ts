export const UNDEF = {} as any;
