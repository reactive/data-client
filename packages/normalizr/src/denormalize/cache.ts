import type { EntityInterface } from '../interface.js';
import { EntityPath } from '../types.js';
import type { INVALID } from './symbol.js';

export default interface Cache {
  getEntity(
    pk: string,
    schema: EntityInterface,
    entity: any,
    computeValue: (localCacheKey: Map<string, any>) => void,
  ): object | undefined | typeof INVALID;
  getResults(
    input: any,
    cachable: boolean,
    computeValue: () => any,
  ): {
    data: any;
    paths: EntityPath[];
  };
}
