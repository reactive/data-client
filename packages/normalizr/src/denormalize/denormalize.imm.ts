import LocalCache from './localCache.js';
import getUnvisit from './unvisit.js';
import type { Schema } from '../interface.js';
import type { DenormalizeNullable } from '../types.js';
import type { INVALID } from './symbol.js';
import { MemoPolicy } from '../memo/Policy.imm.js';

export function denormalize<S extends Schema>(
  schema: S | undefined,
  input: any,
  entities: any,
  args: readonly any[] = [],
): DenormalizeNullable<S> | typeof INVALID {
  // undefined means don't do anything
  if (schema === undefined || input === undefined) {
    return input as any;
  }

  return getUnvisit(
    MemoPolicy.getEntities(entities),
    new LocalCache(),
    args,
  )(schema, input).data;
}
