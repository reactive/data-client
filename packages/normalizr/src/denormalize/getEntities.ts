import type { EntityPath } from '../interface.js';
import type { GetDependency } from '../memo/WeakDependencyMap.js';
import { isImmutable } from '../schemas/ImmutableUtils.js';

export function getEntities<K extends object>(
  state: State<K>,
): GetDependency<EntityPath, K> {
  const entityIsImmutable = isImmutable(state);

  if (entityIsImmutable) {
    return (path: EntityPath) => state.getIn(path);
  } else {
    return ([key, pk]: EntityPath) => state[key]?.[pk];
  }
}

export type State<K extends object> =
  | Record<string, Record<string, K>>
  | { getIn(path: [string, string]): K };

export type GetEntity = GetDependency<EntityPath>;
