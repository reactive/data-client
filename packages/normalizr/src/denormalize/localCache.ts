import type Cache from './cache.js';
import type { EntityInterface } from '../interface.js';
import type { EntityPath } from '../types.js';
import type { INVALID } from './symbol.js';

export default class LocalCache implements Cache {
  private localCache = new Map<string, Map<string, any>>();

  getEntity(
    pk: string,
    schema: EntityInterface,
    entity: any,
    computeValue: (localCacheKey: Map<string, any>) => void,
  ): object | undefined | typeof INVALID {
    const key = schema.key;
    if (!this.localCache.has(key)) {
      this.localCache.set(key, new Map<string, any>());
    }
    const localCacheKey = this.localCache.get(key) as Map<string, any>;

    if (!localCacheKey.get(pk)) {
      computeValue(localCacheKey);
    }
    return localCacheKey.get(pk);
  }

  getResults(
    input: any,
    cachable: boolean,
    computeValue: () => any,
  ): {
    data: any;
    paths: EntityPath[];
  } {
    return { data: computeValue(), paths: [] };
  }
}
