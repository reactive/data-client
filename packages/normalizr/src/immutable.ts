export { DelegateImmutable } from './memo/Delegate.immutable.js';
