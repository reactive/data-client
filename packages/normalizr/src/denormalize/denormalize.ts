import { getEntities } from './getEntities.js';
import LocalCache from './localCache.js';
import getUnvisit from './unvisit.js';
import type { Schema } from '../interface.js';
import { isImmutable } from '../schemas/ImmutableUtils.js';
import type { DenormalizeNullable } from '../types.js';
import type { INVALID } from './symbol.js';

export function denormalize<S extends Schema>(
  schema: S | undefined,
  input: any,
  entities: any,
  args: readonly any[] = [],
): DenormalizeNullable<S> | typeof INVALID {
  // undefined means don't do anything
  if (schema === undefined || input === undefined) {
    return input as any;
  }

  return getUnvisit(
    getEntities(entities),
    new LocalCache(),
    args,
    isImmutable(entities),
  )(schema, input).data;
}
