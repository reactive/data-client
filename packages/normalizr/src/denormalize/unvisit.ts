import type Cache from './cache.js';
import { type GetEntity } from './getEntities.js';
import { INVALID } from './symbol.js';
import { UNDEF } from './UNDEF.js';
import type { EntityInterface } from '../interface.js';
import { isEntity } from '../isEntity.js';
import { denormalize as arrayDenormalize } from '../schemas/Array.js';
import { denormalize as objectDenormalize } from '../schemas/Object.js';
import type { EntityPath } from '../types.js';

const getUnvisitEntity = (
  getEntity: GetEntity,
  cache: Cache,
  args: readonly any[],
  isImmutable: boolean,
  unvisit: (schema: any, input: any) => any,
) => {
  return function unvisitEntity(
    schema: EntityInterface,
    entityOrId: Record<string, any> | string,
  ): object | undefined | typeof INVALID {
    const inputIsId = typeof entityOrId !== 'object';
    const entity =
      inputIsId ? getEntity({ key: schema.key, pk: entityOrId }) : entityOrId;
    if (typeof entity === 'symbol') {
      return schema.denormalize(entity, args, unvisit);
    }

    if (
      entity === undefined &&
      inputIsId &&
      // entityOrId cannot be undefined literal as this function wouldn't be called in that case
      // however the blank strings can still occur
      entityOrId !== '' &&
      entityOrId !== 'undefined'
    ) {
      // we cannot perform WeakMap lookups with `undefined`, so we use a special object to represent undefined
      // we're actually using this call to ensure we update the cache if a nested schema changes from `undefined`
      // this is because cache.getEntity adds this key,pk as a dependency of anything it is nested under
      return cache.getEntity(entityOrId, schema, UNDEF, localCacheKey => {
        localCacheKey.set(entityOrId, undefined);
      });
    }

    if (typeof entity !== 'object' || entity === null) {
      return entity as any;
    }

    let pk: string | number | undefined =
      inputIsId ? entityOrId : (
        (schema.pk(entity, undefined, undefined, args) as any)
      );

    // if we can't generate a working pk we cannot do cache lookups properly,
    // so simply denormalize without caching
    if (pk === undefined || pk === '' || pk === 'undefined') {
      return noCacheGetEntity(localCacheKey =>
        unvisitEntityObject(schema, entity, '', localCacheKey, args, unvisit),
      );
    }

    // just an optimization to make all cache usages of pk monomorphic
    if (typeof pk !== 'string') pk = `${pk}`;

    // last function computes if it is not in any caches
    return cache.getEntity(pk, schema, entity, localCacheKey =>
      unvisitEntityObject(schema, entity, pk, localCacheKey, args, unvisit),
    );
  };
};

function unvisitEntityObject(
  schema: EntityInterface<any>,
  entity: object,
  pk: string,
  localCacheKey: Map<string, any>,
  args: readonly any[],
  unvisit: (schema: any, input: any) => any,
): void {
  const entityCopy = schema.createIfValid(entity);

  if (entityCopy === undefined) {
    // undefined indicates we should suspense (perhaps failed validation)
    localCacheKey.set(pk, INVALID);
  } else {
    // set before we recurse to prevent cycles causing infinite loops
    localCacheKey.set(pk, entityCopy);
    // we still need to set in case denormalize recursively finds INVALID
    localCacheKey.set(pk, schema.denormalize(entityCopy, args, unvisit));
  }
}

function noCacheGetEntity(
  computeValue: (localCacheKey: Map<string, any>) => void,
): object | undefined | typeof INVALID {
  const localCacheKey = new Map<string, any>();
  computeValue(localCacheKey);

  return localCacheKey.get('');
}

const getUnvisit = (
  getEntity: GetEntity,
  cache: Cache,
  args: readonly any[],
  isImmutable: boolean,
) => {
  // we don't inline this as making this function too big inhibits v8's JIT
  const unvisitEntity = getUnvisitEntity(
    getEntity,
    cache,
    args,
    isImmutable,
    unvisit,
  );
  function unvisit(schema: any, input: any): any {
    if (!schema) return input;

    if (input === null || input === undefined) {
      return input;
    }

    if (typeof schema.denormalize !== 'function') {
      // deserialize fields (like Temporal.Instant)
      if (typeof schema === 'function') {
        return schema(input);
      }

      // shorthand for object, array
      if (typeof schema === 'object') {
        const method =
          Array.isArray(schema) ? arrayDenormalize : objectDenormalize;
        return method(schema, input, args, unvisit);
      }
    } else {
      if (isEntity(schema)) {
        return unvisitEntity(schema, input);
      }

      return schema.denormalize(input, args, unvisit);
    }

    return input;
  }

  return (schema: any, input: any): { data: any; paths: EntityPath[] } => {
    // in the case where WeakMap cannot be used
    // this test ensures null is properly excluded from WeakMap
    const cachable = Object(input) === input && Object(schema) === schema;
    return cache.getResults(input, cachable, () => unvisit(schema, input));
  };
};
export default getUnvisit;
