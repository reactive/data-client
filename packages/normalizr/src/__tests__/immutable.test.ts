// eslint-env jest
import { Entity, schema } from '@data-client/endpoint';
import { fromJS, Record, Map } from 'immutable';

import { normalize } from '..';
import { denormalize } from '../denormalize/denormalize';
import { INVALID } from '../denormalize/symbol';
import MemoCached from '../memo/MemoCache';

export function fromJSEntities(entities: {
  [k: string]: { [k: string]: any };
}) {
  return Map(entities).map(v => Map(v));
}

class IDEntity extends Entity {
  id = '';
  pk(parent, key) {
    return this.id || key;
  }
}

class Tacos extends IDEntity {
  type = '';
}

let dateSpy;
beforeAll(() => {
  dateSpy = jest
    .spyOn(global.Date, 'now')
    .mockImplementation(() => new Date('2019-05-14T11:01:58.135Z').valueOf());
});
afterAll(() => {
  dateSpy.mockRestore();
});

describe('immutableJS', () => {
  class Food extends schema.EntityMixin(
    class {
      id = '';
    },
  ) {}
  class MenuData {
    id = '';
    food = Food.fromJS();
  }
  class Menu extends schema.EntityMixin(MenuData, {
    schema: { food: Food },
  }) {}

  test('can normalize', () => {
    const state = fromJS({
      entities: { Food: {}, Menu: {} },
    });
    const v = normalize(Menu, {
      id: '1',
      food: { id: '1', extra: 'hi' },
    });
    // console.log(v);
    const entities = state.mergeDeep(fromJS(v.entities));
    // console.log(entities.getIn(['Menu', '1']));

    // expect(denormalize(Menu, '1', entities)).toMatchSnapshot();
    // expect(denormalize(Menu, '1', fromJS(entities))).toMatchSnapshot();
  });

  test('can denormalize already partially denormalized data', () => {
    const entities = Map({
      Menu: Map({
        1: { id: '1', food: { id: '1' } },
      }),
      Food: Map({
        // TODO: BREAKING CHANGE: Update this to use main entity and only return nested as 'fallback' in case main entity is not set
        1: { id: '1', extra: 'hi' },
      }),
    });

    expect(denormalize(Menu, '1', entities)).toMatchSnapshot();
  });

  test('denormalizes deep entities with records', () => {
    const Food = Record<{ id: null | string }>({ id: null });
    const MenuR = Record<{ id: null | string; food: null | string }>({
      id: null,
      food: null,
    });

    const entities = {
      Menu: {
        '1': new MenuR({ id: '1', food: '1' }),
        '2': new MenuR({ id: '2' }),
      },
      Food: {
        '1': new Food({ id: '1' }),
      },
    };

    expect(denormalize(Menu, '1', fromJSEntities(entities))).toMatchSnapshot();

    expect(denormalize(Menu, '2', fromJSEntities(entities))).toMatchSnapshot();
  });
});
