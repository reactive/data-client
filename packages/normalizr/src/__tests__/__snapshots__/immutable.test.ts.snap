// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`immutableJS can denormalize already partially denormalized data 1`] = `
Menu {
  "food": Food {
    "id": "1",
  },
  "id": "1",
}
`;

exports[`immutableJS denormalizes deep entities with records 1`] = `
Menu {
  "__ownerID": undefined,
  "_values": Immutable.List [
    "1",
    "1",
  ],
  "food": Food {
    "id": "",
  },
  "id": "",
}
`;

exports[`immutableJS denormalizes deep entities with records 2`] = `
Menu {
  "__ownerID": undefined,
  "_values": Immutable.List [
    "2",
    undefined,
  ],
  "food": Food {
    "id": "",
  },
  "id": "",
}
`;
